### PR TITLE
added a method to fetch data from url 

### PR DESCRIPTION
**Pull request by arthurhero**
Added a class method to fetch data from url using criteria, urername and password in GADPerson class

Also fixed a typo in GADStudent class

### CODE REVIEW

File: DB/GADPerson.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
 typedef enum {Student, FacStaff, SGA} GADPersonType;
```
Comment (alexthemitchell) [inline DB/GADPerson.h]: Nit: Errant Newlines (x2)

File: DB/GADPerson.h [hunk 2]
```diff
@@ -30,7 +31,24 @@ typedef enum {Student, FacStaff, SGA} GADPersonType;
 
 -(GADPerson *)initWithDictionary: (NSDictionary *) dict;
 
-+(NSArray <GADPerson*>*)dummyPeople; // returns a array of people, including student, faculty and SGA
++(NSArray <GADPerson*>*)dummyPeople; // returns an array of people, including student, faculty and SGA
+
++ (void) fetchPersonInfoWithCriteria:(NSDictionary*)crit andUsername:(NSString*)usrname andPassword:(NSString*) pw completionHandler:(void(^_Nonnull)(NSArray<GADPerson *> *))completion; // This is important!!!!!!!!
+/*
```
Comment (alexthemitchell) [inline DB/GADPerson.h]: I love your documentation! Can you format it as described [here](http://www.appcoda.com/documenting-source-code-in-xcode/)? That way, we'll have some documentation tooltips in Xcode! I'll show them off on Thursday.

File: DB/GADPerson.h [hunk 3]
```diff
@@ -30,7 +31,24 @@ typedef enum {Student, FacStaff, SGA} GADPersonType;
 
 -(GADPerson *)initWithDictionary: (NSDictionary *) dict;
 
-+(NSArray <GADPerson*>*)dummyPeople; // returns a array of people, including student, faculty and SGA
++(NSArray <GADPerson*>*)dummyPeople; // returns an array of people, including student, faculty and SGA
+
++ (void) fetchPersonInfoWithCriteria:(NSDictionary*)crit andUsername:(NSString*)usrname andPassword:(NSString*) pw completionHandler:(void(^_Nonnull)(NSArray<GADPerson *> *))completion; // This is important!!!!!!!!
```
Comment (alexthemitchell) [inline DB/GADPerson.h]: `andUsername` -> `Username`
`usrname` -> `username`
`andPassword` -> `Password`
`pw` -> `password`

I'll post a link to a styleguide to enumerate conventions and best practices soon.

Also, we want to keep all lines less than 100 characters long. When we use newlines, we want to align on the colons, as below:
```
+ (void) searchWithCriteria:(NSDictionary *)criteria 
                   Username:(NSString *)username 
                   Password:(NSString *) password
          completionHandler:(void(^_Nonnull)(NSArray<GADPerson *> *))completion;
```
```

File: DB/GADPerson.m [hunk 4]
```diff
@@ -3,6 +3,12 @@
 #import "GADSGA.h"
 #import "GADFacStaff.h"
 
+@interface GADPerson()
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: This entire interface section is unnecessary. Remove it.

File: DB/GADPerson.m [hunk 5]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
     return people;
 }
 
++ (void) fetchPersonInfoWithCriteria:(NSDictionary*)crit andUsername:(NSString*)usrname andPassword:(NSString*) pw completionHandler:(void(^_Nonnull)(NSArray<GADPerson *> *))completion {
+
+    NSMutableArray * queryItems = [NSMutableArray array];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Let's use generics! Try `NSMutableArray<NSURLQueryItem *> *queryItems = [NSMutableArray<NSURLQueryItem *> new];`

This way, we have compile-time typechecking for the array. It prevents bugs!

File: DB/GADPerson.m [hunk 6]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    for (NSString * key in crit){
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: No space between `*` and variable name. Use `NSString *key in crit`. Here and throughout.

File: DB/GADPerson.m [hunk 7]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+        [queryItems addObject:[NSURLQueryItem queryItemWithName:key value:crit[key]]];
+    }
+    NSURLComponents *components = [NSURLComponents componentsWithString:@"https://itwebappstest.grinnell.edu/DotNet/WebServices/api/db"];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Can we declare this as a constant at the top of the file? `static const NSString *apiURL = @"...";` and then `NSURLComponents *components = [NSURLComponenets componentsWithString:apiURL];` here.

File: DB/GADPerson.m [hunk 8]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    components.queryItems=queryItems;
+    NSURL *url = components.URL;
+    
+    NSString *Post = [NSString stringWithFormat:@"{'un':'%@', 'pw':'%@'}",usrname,pw];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: We use camelcase with a leading lowercase for variable names. Could we call this the `authenticationString` instead?

File: DB/GADPerson.m [hunk 9]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    NSData *PostData = [Post dataUsingEncoding:NSUTF8StringEncoding];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: `bodyData`

File: DB/GADPerson.m [hunk 10]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    NSString *PostLength = [NSString stringWithFormat:@"%lu",(unsigned long)[PostData length]];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: `bodyLength`

File: DB/GADPerson.m [hunk 11]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    
+    NSMutableURLRequest *req = [NSMutableURLRequest requestWithURL:url cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60.0];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Can we make the timeout a constant at the top too? I forget what time `timeoutInterval` takes, but we'd declare `const <type> timeoutInterval = 60.0;` up with the `apiURL`.

File: DB/GADPerson.m [hunk 12]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    [req setHTTPMethod:@"POST"];
+    [req setValue:PostLength forHTTPHeaderField:@"Content-Length"];
+    [req setValue:@"application/json" forHTTPHeaderField: @"Content-Type"];
+    [req setHTTPBody:PostData];
+    
+    NSURLSessionDataTask *task = [[NSURLSession sharedSession]dataTaskWithRequest:req completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (error) {
+            NSLog(@"EEERRRRORRRRR");
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: I'm okay with leaving this here, but can you make a Github Issue to add error handling  here? :D

File: DB/GADPerson.m [hunk 13]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+            return;
+        }
+        
+        NSMutableArray <GADPerson*> * people=[NSMutableArray array];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: `[NSMutableArray<GADPerson *> new]`.

File: DB/GADPerson.m [hunk 14]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+        NSArray *arr = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Add error handling here to that Github issue (above)

File: DB/GADPerson.m [hunk 15]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+        
+        for (NSDictionary * entry in arr){
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: No space before `entry`

File: DB/GADPerson.m [hunk 16]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+            
+        GADPerson *person = [[GADPerson alloc]initWithDictionary:entry];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Space before `initWithDictionary`

File: DB/GADPerson.m [hunk 17]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+         
+         [people addObject:person];
+         }
+        
+        completion(people);
+
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Remove these newlines

File: DB/GADPerson.m [hunk 18]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+         
+         [people addObject:person];
+         }
+        
+        completion(people);
+        
+    }];
+    
+    [task resume];
+
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Remove this newline

File: DB/GADPerson.m [hunk 19]
```diff
@@ -95,6 +101,50 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+         }
+        
+        completion(people);
+        
+    }];
+    
+    [task resume];
+    
+}
+
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Remove this newline :D

**Review by alexthemitchell — CHANGES_REQUESTED** [review]
This is great, Ziwen! Just a few stylistic changes and we can get this merged!

File: DB/GADPerson.m (hunk 12 above)
Comment (arthurhero) [reply DB/GADPerson.m]: Sure. But who should I assign it to?

File: DB/GADPerson.m (hunk 19 above)
Comment (arthurhero) [reply DB/GADPerson.m]: OK, lol

**Comment by arthurhero** [comment]
I changed a line in the initWithDictionary method in GADPerson class from
`[dict valueForKey:@"office_email"] isEqual: @""]`
to
`[dict valueForKey:@"office_email"] == (id)[NSNull null]`
since the latter one seems to be the correct way to test null value returned by a JSON string.

File: DB/GADPerson.m [hunk 20]
```diff
@@ -89,14 +92,49 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
     person3.office_email=@"sgaprez";
     person3.office_hours=@[@"Sunday: 7-8pm Grille, 8-9pm JRC 222",@"  Monday: 9 - 3 Commons",@"  Tuesday: 7 -7 JRC Main",@"  Wednesday: 12 - 12 Raynard's Office"];
 
-    
     people=@[person1,person2,person3];
     
     return people;
 }
 
++ (void) fetchPersonInfoWithCriteria:(NSDictionary*)criteria Username:(NSString*)usrename Password:(NSString*) password completionHandler:(void(^_Nonnull)(NSArray<GADPerson *> *))completion {
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: Can you make it look pretty here too? Align on the colons!

Comment (alexthemitchell) [reply DB/GADPerson.m]: Also, `username` is misspelled here and throughout :D

File: DB/GADPerson.m [hunk 21]
```diff
@@ -89,14 +92,49 @@ -(GADPerson *)initWithDictionary: (NSDictionary *) dict {
...
+    [req setValue:@"application/json" forHTTPHeaderField: @"Content-Type"];
+    [req setHTTPBody:bodyData];
+    
+    NSURLSessionDataTask *task = [[NSURLSession sharedSession]dataTaskWithRequest:req completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (error) {
+            NSLog(@"EEERRRRORRRRR");
+            return;
+        }
+        NSMutableArray <GADPerson*> *people=[NSMutableArray<GADPerson*> new];
+        NSError *JSONParsingError = [[NSError alloc]initWithDomain:NSCocoaErrorDomain code: NSPropertyListReadCorruptError userInfo:nil];
```
Comment (alexthemitchell) [inline DB/GADPerson.m]: You don't actually need to set this to anything. Just `NSError *JSONParsingError;` here and the rest should work as expected.

**Review by alexthemitchell — APPROVED** [review]
This is fantastic! Please make these three small changes and then merge to master when you're ready.

**Pull request merged**